### ISSUE TITLE
[DX] Use region names as classes in flexible templates

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#5032]
**Description of the need**

Inspired by https://forum.backdropcms.org/forum/why-were-all-handy-classes-removed, I think the region class in the template should include the region name. Currently regions only include `l-col col-md-NUM` where NUM is the number of columns.

Currently it's possible to add classes to rows in the flexible layout UI, but not for regions. However, we can add region names. I don't think there's any drawback to just **turning the region name into a class and then printing it in the template**.

Markup for "Region 1"  before PR: `<div class="l-col col-md-8">`
Markup for "Region 1" after PR: `<div class="l-col col-md-8 l-region--region-1">`

---

Advocate: @jenlampton 

---

PR by @herbdool: https://github.com/backdrop/backdrop/pull/3591

Alt approach by @herbdool https://github.com/backdrop/backdrop-issues/issues/5044 - adds an option for custom classes on the region.

**Comment by olafgrabienski** [comment]
Many thanks for the PR, @herbdool! I've tested it in the sandbox website, and it works as expected. For example, the output of a custom region called "Bottom 1" is `<div class="l-col col-md-4 bottom-1">`.

**Comment by klonos** [comment]
I've tested this as well, and confirm that it works as expected 👍🏼 ...the code change is minimal and straight-forward, but should the classes have the prefix `region-` for consistency?

**Comment by herbdool** [comment]
@klonos thx. In the core templates like https://github.com/backdrop/backdrop/blob/1.x/core/layouts/moscone/layout--moscone.tpl.php the format is like `l-sidebar` for the region. Maybe that should be the prefix instead?

**Comment by klonos** [comment]
> Maybe that should be the prefix instead?

Yup, or that 👍🏼 ...I'd prefer it if someone else that does theming a lot chimed in though. ... @jenlampton ... @stpaultim? ...others?

**Comment by indigoxela** [comment]
I'm happy with both prefixes, `l-` or `region-`. My personal preference would be the less ambiguous `region-`, because it's much easier to understand when reading it in a CSS file.

**Comment by olafgrabienski** [comment]
> `region-` (...) overlap with pre-existing classes is less probable.

Good point! I've had another look at the PR's sandbox. Using the `l-` prefix, we would end up with the same `l-header` class in line 1 and 4 of the following example:

```
1 <header class="flexible-row (...) l-header">
2   <div class="container (...) region_12">
3     <div class="l-flexible-row row">
4       <div class="l-col col-md-12 l-header">
```

Better with `region-` prefix:

```
1 <header class="flexible-row (...) l-header">
2   <div class="container (...) region_12">
3     <div class="l-flexible-row row">
4       <div class="l-col col-md-12 region-header">
```

**Comment by herbdool** [comment]
Now I'm wondering if it should be just a separate field for additional classes so people can make their own decision. This forum post made me think that my pr doesn't solve their issue: https://forum.backdropcms.org/forum/using-bootstrap-classes-flexible-layouts-regions.

Then people can decide the class name including the prefix or no prefix, or other classes like bootstrap ones.

**Comment by indigoxela** [comment]
> Now I'm wondering if it should be just a separate field for additional classes so people can make their own decision.

I see some value in automatic region classes by region name. IMO it's more convenient for site builders than adding them manually all over the place in all layouts. As long as we take care, that there's no class name overlap, everything should be fine.

Re the forum post @herbdool linked: the problem described there is slightly different, so their problem wouldn't be solved, I guess.

But this one possibly: https://forum.backdropcms.org/forum/why-were-all-handy-classes-removed

**Comment by jenlampton** [comment]
I also like automatic classes. Core uses dashes and not underscores to
separate words in class names though.

**Comment by olafgrabienski** [comment]
> Core uses dashes and not underscores to separate words in class names though.

@jenlampton Do you refer to `region_12` from the example above (`<div class="container (...) region_12">)`? IIRC, this class is provided by core in Flexible layouts ;-)

**Comment by klonos** [comment]
> I also like automatic classes.

Me too 👍🏼

**Comment by jenlampton** [comment]
> Do you refer to region_12 from the example above (<div class="container (...) region_12">)?

I did!

> IIRC, this class is provided by core in Flexible layouts ;-)

ohnoes, I thought that was from the latest PR :( well, we can add the correct class anytime, but we can't remove those until 2.x :D

**Comment by indigoxela** [comment]
There seems to be general support for the automatic classes based on region names.

So all this PR needs is a prefix to make the names less ambiguous and it would be ready?

The need for additional UI to get more influence on the (wrapping?) classes can be a follow-up IMO. @herbdool what do you think?

**Comment by herbdool** [comment]
I've got two PRs on this issue now. The first adds "region-REGIONNAME" as an automatic class. The second goes a bit further by adding a new "additional classes" field per region. I think the second would be a great addition if it doesn't get bogged down in controversy. If it does let's just merge the first and I'll create a new issue for the new field.

**Comment by olafgrabienski** [comment]
I've tested both PRs using their sandbox websites, both work for me.

The results from a flexible layout template with two custom regions (Bottom 1 and Bottom 2):

PR 3581 (automatic classes)
`<div class="l-col col-md-6 region-bottom-1">`
`<div class="l-col col-md-6 region-bottom-2">`

PR 3584 (automatic + custom classes)
`<div class="l-col region-bottom-1 col-md-6 bg-orange">`
`<div class="l-col region-bottom-2 col-md-6 bg-blue">`

While I like the option to add custom classes, I'm not sure about the configuration form with nested collapsible fieldsets (collapsed by default) of PR 3584 – it looks quite busy. I'd suggest to use not collapsible fieldsets, or div elements, for the inner single region fields. (And maybe rename the outer `Regions` legend to `Region names and classes`).

See a screenshot comparison of both configuration forms:

**PR 3581 (automatic classes)**
<img width="450" alt="backdrop-pr-3581-layout-region" src="https://user-images.githubusercontent.com/14188307/113936868-b4723800-97f8-11eb-9031-e42eacfdba51.png">

**PR 3584 (plus custom classes)**
<img width="450" alt="backdrop-pr-3584-layout-region" src="https://user-images.githubusercontent.com/14188307/113936872-b5a36500-97f8-11eb-8cb4-819beee8126b.png">

**Comment by jenlampton** [comment]
Question: could changing a regions's name break your site? (would the class change, too?) 

If so, this still needs work. (thinking... Maybe the machine name of a region shouldn't change once it's created?)

**Comment by herbdool** [comment]
Or maybe just put a message that changing the name will change the class? I had first looked into using the machine name form type but it requires unique names which we don't need here (I think).

**Comment by jenlampton** [comment]
Unique names might be good? Otherwise, how would you get your CSS to target one "stripe" region differently from another?

**Comment by herbdool** [comment]
Currently regions in flexible templates aren't unique at all. Are most people going to need unique automatic classes based on the region name? If we also include the additional classes then there's still more options on making them unique.

Plus there are already lots of ways for someone to mangle their site layout by changing rows and columns. How much of an issue is it going to be? Though maybe the difference is that it needs to clear that the region name will be used as a class.

**Comment by indigoxela** [comment]
> Unique names might be good?

Not necessarily. If someone *does* want to apply the same styles to different regions, they couldn't. Unlike CSS `id` attributes, classes are supposed to get re-used.

**Comment by olafgrabienski** [comment]
> put a message that changing the name will change the class? I had first looked into using the machine name form type but it requires unique names which we don't need here (I think).

A message (changing the name changes the class as well) sounds reasonable to me. People who actually use these classes in their CSS, will be okay to change the CSS then. And I agree, we shouldn't force unique names.

> Unique names might be good? Otherwise, how would you get your CSS to target one "stripe" region differently from another?

There are various ways: custom classes if we go with PR 3584. Otherwise using `:nth-of-type` or `:nth-child` selectors, for instance. Also, I *think*, most of the times a "stripe" layout wouldn't be made of these regions but of rows or row wrappers.

**Comment by jenlampton** [comment]
This isn't an interface for setting class names though, It's an interface for setting human-readable labels. We're should be trying to achieve a sensible default in the automatic-classes, and leave anything more complicated to contrib.

> If someone does want to apply the same styles to different regions, they couldn't. 

It's easy to apply the same styles to different selectors:

```
.selector1,
.selector2 {
```

It's hard to apply different styles to the same selector:

```
.selector {}
.selector:nth-child(n+2) {}
```

Backdrop should require the lowest level of proficiency possible for all features to keep the software easy to use, so I believe a sensible default here would be unique classes for each region.

> A message (changing the name changes the class as well) sounds reasonable to me. 

People aren't going to change the human-readable labels of their regions to try to achieve some fancy class naming, and I don't think we should clutter the UI with description text telling them classes will be created. We don't do that for blocks, menus, or any other element. The sensible defaults are just there.

If people do want to control the classes that are added, they'll use a contrib module that provides a UI for classes (or get into the templates in the theme).

> Unlike CSS id attributes, classes are supposed to get re-used.

Maybe we should use the unique region name to create an ID then instead of a class? We've removed most IDs from core because they are really hard to work with in terms of specificity, but here they could also solve the problem of how to build a link to a particular part of the page.

**Comment by indigoxela** [comment]
I did a quick test in the sandbox - I like it! :+1: 

One small thing: the automatic prefix maybe doesn't make sense anymore. People can edit that value, so they can add any "prefix" they want to prevent overlap. In my quick test I ended up with a class name like "region-region-content" while the value I could edit was only "region-content". I had no influence on the prefix like `'region-' . $machine_name;`, which felt a bit odd.

Another thought:

> Even though there's a callback function to check for uniqueness it's basically accepting anything.

So why bothering anyway? What's a function good for, that just always returns "FALSE" and technically wouldn't be able to guarantee uniqueness anyway?

**Comment by olafgrabienski** [comment]
> the automatic prefix maybe doesn't make sense anymore. People can edit that value, so they can add any "prefix" they want

Or let's use the field creation pattern, where the `field_` prefix exists but isn't editable:

<img width="411" alt="backdrop-field-edit-machine-name" src="https://user-images.githubusercontent.com/14188307/114143555-9a2d7c80-9914-11eb-9990-36b2d0759624.png">

**Comment by olafgrabienski** [comment]
> Maybe we should use the unique region name to create an ID then instead of a class? We've removed most IDs from core because they are really hard to work with in terms of specificity, but here they could also solve the problem of how to build a link to a particular part of the page.

We're already going to allow custom IDs to build links to particular parts of a page in https://github.com/backdrop/backdrop-issues/issues/4052, namely for the *outer* rows of a flexible layout (`flexible-row`). I think, there it makes sense to have IDs [as an exception](https://github.com/backdrop/backdrop-issues/issues/4052#issuecomment-533705972), but not for the regions which are quite deep inside of rows.

**Comment by herbdool** [comment]
Good suggestion @olafgrabienski. I've added the "region-" prefix in the UI.

@indigoxela since people decided it should be a machine_name, it meant we had to have some kind of callback function for checking uniqueness. It's not optional. https://docs.backdropcms.org/form_api#machine_name. And it definitely has to be a real function - it won't accept something like `'exists' => FALSE,`

**Comment by indigoxela** [comment]
As we have a dedicated text field for additional classes now, maybe the machine_name field makes no sense anymore. Two fields are doing the same job. Why two of them and different ones, anyway (machine_name and text)? I think, that's the wrong direction.

**Comment by herbdool** [comment]
From what I understood, people liked the automatic part of the one class but there is also a need for additional classes. The additional classes is so people can add generic classes like `bg-red` or `col-sm-2`, which would be hard otherwise.

**Comment by jenlampton** [comment]
I still believe the UI for classes is overkill, and belongs in contrib.

**Comment by olafgrabienski** [comment]
> From what I understood, people liked the automatic part of the one class but there is also a need for additional classes.

The problem seems to be that *some* people prefer the automatic part, while *others* see more value in additional classes. Maybe it's time to come back to [this introduction](https://github.com/backdrop/backdrop-issues/issues/5032#issuecomment-815089528) of the second PR:

> I think [PR 3584] would be a great addition if it doesn't get bogged down in controversy. If it does let's just merge the first [PR 3581] and I'll create a new issue for the new field.

I guess I'd prefer a new issue for additional classes over a contrib solution. If the latter, I'd prefer it would be part of sort of a Flexible layout extra module which resolves more issues than just classes for layout regions.

**Comment by jenlampton** [comment]
I don't think there are any objections to the automatic ones, though? So
maybe we move that through first (as this issue / a task) and get it into a
bugfix release.

Then if people still want a core UI we can open a proper feature request,
add automated tests, and get it added to the next minor release milestone.

I expect once people have the classes they need, the desire for a UI will
decrease, and contrib will seem more like the correct place for that sort
of control.

**Comment by klonos** [comment]
I agree with what @jenlampton said ^^ 👍🏼

**Comment by herbdool** [comment]
Actually the person in the forum also had the problem of applying generic classes to a region. See https://forum.backdropcms.org/forum/using-bootstrap-classes-flexible-layouts-regions @klonos

**Comment by jenlampton** [comment]
I still think that use-case is better solved by contrib. I'm not sure we have enough feedback to be sure that's an 80% usecase.

**Comment by indigoxela** [comment]
@herbdool There are two forum topic you're trying to address, but actually the one this issue tries to handle is only one of them.

The one we quickly found consensus on is the "automatic class by region name" topic. The second one (UI for additional classes) needs for sure more discussion and should - in my opinion - get handled in a follow-up issue. It shouldn't be a blocker for the easier one.

**Comment by herbdool** [comment]
@indigoxela yes, I've been open about me wanting to address both items here. They involve very similar problems and solutions. So I felt it made sense to try to sus out a way to address them both.

At this point, other than being busy, I've been wondering if there's more feedback from others. I've asked the forum poster as well. I want to know more before doing more changes, including splitting this issue.

It's not clear to me why @jenlampton sees the one approach (automatic class) as an "80% solution" but not the other. We don't have data on either and it seems to me that it might be partly based on different preferences for using CSS: writing CSS components versus utility classes. (I'm no CSS expert but this seems to be two distinct approaches.) Only the latter PR allows someone to use utility classes.

I'm also hesitant to spilt the issue just yet because it's not clear how opposed some people are you this. It sounds like even if this was put into a different pr that it would get rejected as needing to a contrib module.

Though I'm not sure if I would want to create a contrib module for such a small bit of functionality. Sounds like too much work. Maybe I'd do it if I had a current project that needed it. (I do, but it's a Drupal 9 project with layout builder).

So we'll see.

**Comment by herbdool** [comment]
By the way I'm not sure about the automatic classes as meeting bug fix release criteria. It's technically new functionality and requires storing new values.

**Comment by indigoxela** [comment]
@herbdool I'm sorry to see, that you seem a bit frustrated - I can totally understand that, though. You spent some time with providing the extended solution, only to realize that there are concerns regarding the UI.

However, I still think that the automatic classes based on region names (with prefix) are a good enhancement, helpful for styling.

> By the way I'm not sure about the automatic classes as meeting bug fix release criteria.

The change is small, the issue is marked as "task", so a bug fix release seems OK - or are there objections regarding that? In practice - if we don't waste too much time with discussion ( :grin: ) - it wouldn't make much difference, as the next minor and bugfix release are likely to come out at the same time.

**Comment by klonos** [comment]
> The change is small, the issue is marked as "task", so a bug fix release seems OK - or are there objections regarding that?

Nope, no objections I believe:
- if we are to add only the automatic classes, then this is a task
- if we are to add the UI that allows people to customize classes, then that's a feature request

**Comment by olafgrabienski** [comment]
> It sounds like even if this was put into a different pr that it would get rejected as needing to a contrib module.
> Though I'm not sure if I would want to create a contrib module for such a small bit of functionality.

Good point! Maybe it's time to discuss (in another issue, or during a weekly meeting) when and how contrib solutions are a good approach, and when it's better to have something in core even if we're not sure if it matches the 80% criteria.

I agree however, the addition of automatic classes shouldn't be blocked by the (so far) more controversial option to add custom classes. I also agree, a contrib module only for custom classes on Flexible layout regions doesn't make sense. It should solve more things, e.g. as part of a "Flexible layout extra" module. Or maybe as part of a module which tries to make working with utility classes in Backdrop generally more easy.

**Comment by jenlampton** [comment]
> Maybe it's time to discuss ... when it's better to have something in core even if we're not sure if it matches the 80% criteria.

The answer to this is "never" -- this is why we have the 80% criteria :)  We've made one exception to this rule so far in Backdrop, and in that case it was because without a change in core, it would have been impossible to do in contrib.

**Comment by laryn** [comment]
I think part of the problem is that we have no data to support whether or not something is 80% so we have a criteria that sounds very precise but is at the moment somewhat subjective.

My 2¢ -- the automatic classes are an improvement but I agree with @herbdool that additional classes would be even better. I'm also in favor of merging the automatic PR and opening a new issue to continue that discussion.

**Comment by herbdool** [comment]
Third PR: https://github.com/backdrop/backdrop/pull/3591.

To recap:

First PR: simplest. Only changes one line in the template to create a region class.
Second PR: adds a machine name form type for setting the region class, plus an additional classes field. Requires changes to store configuration. Has a test.
Third PR: adds a machine name form type for setting the region class. Requires changes to store configuration and an upgrade hook. Has a test.

The first PR is already off the table as per discussion above.

Only the first PR could possibly be labelled a "task" and be part of a bug fix, IMHO. The other two are feature requests and require upgrade hooks. So the third PR would still require waiting until the next minor release.

Looks like there's agreement for opening a new issue for the "additional classes", even by people support it so I'll do that. But I wanted to add more thoughts on the idea of the "80% threshold".

Like @laryn said, we don't have data on the 80%. In a lot of cases it would be hard to get this data anyway, particularly for small feature requests like adding one field "Additional CSS classes".

I'm not certain we could even claim that the flexible layout templates meets this threshold. It *is* user friendly, but will most people need it? I don't know - I've only used it once.

And what about the additional row classes field? Does that pass the 80% threshold?

![Screenshot from 2021-04-16 14-54-09](https://user-images.githubusercontent.com/1071483/115071732-9b3a4d00-9ec4-11eb-87c7-0d50e451c27d.png)

Or how about the "Additional CSS classes" for blocks in layouts? Or all the way to set custom CSS classes in Views?

It seems pretty arbitrary and subjective to draw a line with regions. And it's pretty hard to avoid that for any of us (including me), so we should be open about this and willing to provide more justification for our position.

**Comment by klonos** [comment]
@herbdool thanks for splitting the other issue 👍🏼

> But I wanted to add more thoughts on the idea of the "80% threshold". ...It seems pretty arbitrary and subjective to draw a line with regions. And it's pretty hard to avoid that for any of us (including me), so we should be open about this and willing to provide more justification for our position.

You'd be happy to know that @olafgrabienski mentioned this during our last dev meeting (thanks for that @olafgrabienski 👍🏼 ), and although we did not have enough time/energy to discuss it at the time, we all agreed to add it to the meeting agenda for next week.

To everyone in this thread, lets please wait for that discussion to happen, and not derail this issue here further with such discussions. This is a matter that deeply interests me, and it is strongly tied to #285 and the whole telemetry initiative.

**Comment by stpaultim** [comment]
I commented on the second PR before coming back and reading this discussion. 

1) It seems like there is consensus on adding an automatic class to the regions and I support that. 
2) I'll comment on the ability to add custom classes in the follow-up thread. 

Seven people so far, @klonos, @laryn, @indigoxela, @herbdool, @olafgrabienski, @jenlampton and myself, have all commented in this discussion. Everyone seems to support the new automatic class.

**Comment by herbdool** [comment]
I fixed the conflict. But I tested and discovered the region names aren't sticking after saving.

**Comment by herbdool** [comment]
Anyone want to do a code review? @laryn @indigoxela @klonos @jenlampton

**Comment by indigoxela** [comment]
@herbdool your branch (still) has conflicts, that need to get resolved.

**Comment by herbdool** [comment]
Thanks @indigoxela. I've resolved the conflicts - it's because there was another fix to the flexible layout template.

At the risk of stirring the pot again, I've updated my alternative PR https://github.com/backdrop/backdrop-issues/issues/5044#issuecomment-984102195 so the CSS classes field for regions isn't visible until checked.

**Comment by stpaultim** [comment]
@herbdool - Is there something specific you want feedback on for this issue? Seems like we have consensus on the functionality of this PR. I just tested it again and think it looks great. 

I also support the change in #5044, but am pretty sure that we need to get this one merged, before we'll get consensus on that one. I am worried that the "needs more feedback" label might slow this down and I'm not sure if it's necessary. Seems like this one is really close to RTBC.

**Comment by indigoxela** [comment]
And this is what we have here in this PR (still works for me):

![region-dialog-2](https://user-images.githubusercontent.com/842066/145356369-c8330911-ce4d-4554-a72f-2bc948830294.png)

Also nice! As the two PRs achieve something very similar in different ways, we probably have to decide.
@herbdool how does it feel to be your own competitor? :smirk:

**Comment by stpaultim** [comment]
> Also nice! As the two PRs achieve something very similar in different ways, we probably have to decide.
@herbdool how does it feel to be your own competitor?

In theory, I support both of these PR's. But, I assume that one will need to be approved before we can decide on the other, because of how they are interrelated. Which is confusing me, because we're making progress on both of them simultaneously and  I'm reluctant to get behind either one, because I'm not sure which one might be approved first and/or how they will impact each other. I'm confident that they can work together. I just don't understand what the plan is. 

To be clear, I'm not not opposed to either and at this point am happy to stand back and see what happens. I just think that both of there are really close to approval, but that might be difficult without some specific proposal on how to manage them. 

@herbdool - Do you have a suggestion for how to proceed?

**Comment by herbdool** [comment]
@stpaultim both PRs could work though I personally think https://github.com/backdrop/backdrop-issues/issues/5044 is more important since it provides more flexibility. This one only allows just one custom CSS class but not much control over what it's called.

**Comment by jenlampton** [comment]
> In the core templates like https://github.com/backdrop/backdrop/blob/1.x/core/layouts/moscone/layout--moscone.tpl.php the format is like `l-sidebar` for the region. Maybe that should be the prefix instead?

The `l-` prefix for classes indicate that those elements are part of the layout. I think the `.l-` pattern would be good to use here as well. They communicate that layout-specific styling can target these elements (without risk of targeting something something that's not part of a layout).

> My personal preference would be the less ambiguous region-, because it's much easier to understand when reading it in a CSS file. And overlap with pre-existing classes is less probable.

I find `region-` more ambiguous, because people won't instantly know it was part of a layout. We could solve both problems by using a class name beginning with `l-region-`. or even `l-region--`, which would better help differentiate between the prefix and the region name (especially sine the region name 

> both PRs could work though I personally think #5044 is more important since it provides more flexibility.

I said this in today's meeting, but I'm going to put my thoughts here for posterity. I think this issue is much more important because without this region-specific class, it's impossible for people to target a specific region (which makes this issue more of a bug, actually).  

If we had this class in core, nobody would need #5044.  Right now, #5044 is very attractive to people because it is also solving this bug -- but at the cost of adding more clutter to an already complex interface. This automatic class solves the bug without the increased complexity.

**Comment by jenlampton** [comment]
I've submitted [a PR](https://github.com/herbdool/backdrop/pull/16) against @herbdool's latest to change the class name to include the double hyphen, and the layout indicator. 

Markup after the PR would look like `<div class="l-col col-md-8 l-region--region-1">`.

**Comment by jenlampton** [comment]
@herbdool I added the bug report label intentionally, because without a region-specific class, it's impossible for people to target a specific region. Did you feel that problem is not really a bug? 

I'm removing the needs work label because the PR was merged (thank you!). Anyone else want to give this another review? If not, this is RTBC from me! :)

**Comment by indigoxela** [comment]
What's a little unclear to me is the fact, if this PR is still needed when we're also adding #5044.

Does it make sense to add both?

**Comment by herbdool** [comment]
Now that there's consensus on #5044 there's less reason for this one, especially since #5044 is also, aside from the custom classes, adding automatically a unique class indentifier such as `l-region--2`.

But if people prefer the automatic class to be the region name we could have both. The UI for this one, though, is a bit of a one-off. I don't we've used the machine name feature anywhere else to generate a class. But it people are ok with that I recommend we merge #5044 first and then I can update this PR from upstream.

**Comment by indigoxela** [comment]
> But if people prefer the automatic class to be the region name we could have both.

TBH my personal favorite is #5044. Setting css classes as I need them, even multiple, seems more useful to me. But that's just my 2c. :grinning:

**Comment by jenlampton** [comment]
Adding the other feature doesn't fix this bug.

It's still important for people to be able to target a specific region with
CSS. There's no reason to make them fix the bug manually in config before
they can complete their task. They may not even have access to the flexible
layouts UI.

**Comment by jenlampton** [comment]
>  I don't we've used the machine name feature anywhere else to generate a class.

At yesterday's meeting, @quicksketch recommended that we switch to using a machine name FAPI element, so that we don't need to worry about transliteration for the class name, should someone name the region using Chinese or Cyrillic characters. (This is what @herbdool was referring to, above.) 

A machine name would give people the freedom to change the Region name, without breaking any CSS identified with a previous region-specific class. @quicksketch also mentioned that having a machine name for a region could be useful for other reasons in the future. 

We do use machine names to generate classes in lots of other places in core (content types, etc) the only difference is that we use those machine names for other things too :)

**Comment by herbdool** [comment]
I've now merged in the functionality from this issue (creating a default class from a machine name) into #5044. I figured it best since we got consensus to merge both and it's easier to combine them before merging.

**Comment by klonos** [comment]
Since this issue was closed I'm removing the labels. Feel free to re-add them if it's reopened.
